Add proposal about auto-approval mechanism for optimization proposals (#56)
strimzi/proposals: 038-optimization-proposal-autoapproval.md

# Auto-approval mechanism for optimization proposals

> This proposal is about allowing the user to create a `KafkaRebalance` custom resource for getting an optimization proposal and running a rebalance, but without the need of the manual approval via the `strimzi.io/rebalance=approve` annotation.

Reviewer: ```suggestion
This proposal is about allowing the user to create a `KafkaRebalance` custom resource for getting an optimization proposal and running a rebalance, but without the need for the manual approval via the `strimzi.io/rebalance=approve` annotation.
```

> It means that the user can just create the `KafkaRebalance` custom resource and the corresponding generated optimization proposal will be automatically approved.

Reviewer: ```suggestion
This means that the user can simply create the `KafkaRebalance` custom resource and the corresponding generated optimization proposal will be approved automatically .
```

## Current situation

> Currently, when the user wants to run a cluster rebalance, he has to create a `KafkaRebalance` custom resource for getting an optimization proposal first.

Reviewer 2: Maybe this way it is better to avoid any he/she discussions ;-)

```suggestion
Currently, when the users want to run a cluster rebalance, they have to create a `KafkaRebalance` custom resource for getting an optimization proposal first.
```

> Currently, when the users want to run a cluster rebalance, they have to create a `KafkaRebalance` custom resource for getting an optimization proposal first.

Reviewer: ```suggestion
Currently, when the users want to run a cluster rebalance, they have to create a `KafkaRebalance` custom resource in order to generate an optimization proposal first.
```

After getting the proposal, the only way to start the rebalancing operation is by approving it by annotating the custom resource with the `strimzi.io/rebalance=approve` annotation.
There is the need for a manual intervention of the user and this is a two steps process.

## Motivation

In some cases, the manual approval process of the rebalance proposal is not necessary.
Automatically approving an optimization proposal can save time from an operational point of view.
It enables more automation where just creating a `KafkaRebalance` custom resource can go straight to the cluster rebalance without any additional manual interaction.

## Proposal

The `KafkaRebalance` custom resource can be extended with a new `spec.autoApproval` field for this purpose.

### Auto-approval

The auto-approval can be enabled just adding an empty `spec.autoApproval` field.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:
    strimzi.io/cluster: my-cluster
spec:
  mode: # any mode

>   autoApproval: {}

Reviewer 3: Do we know that (in the future when this is acting as more than an unconditional "approve it automatically" flag) it would actually be an _object_, rather than a _list_ of criteria to be met? That is, I could imagine that in that future we'd actually want something like

```yaml
spec:
  mode: ...
  approvalCriteria:
    - criterionType: Foo
      threshold: 100
    - criterionType: Bar
      someCriterionParameter: true
```

I _guess_ having in intermediate object might allow some flexibility (e.g. you could have an `anyOf` property and an `allOf` property to provide different ways of combining multiple criteria), but it's equally not clear to me that we'd need such flexibility (and hence no need for an intermediate object either).

Reviewer 3: Note that `approvalCriteria` could allow use to handle the existing manual approval in a more consistent way. I.e. rather than using an annotation we could just say the API was to update the `spec.approvalCriteria` with a criterion that always evaluated as true. For example:

```yaml
spec:
  mode: ...
  # approvalCriteria: abence of property means not approved
```

And this CR

```yaml
spec:
  mode: ...
  approvalCriteria: 
    - criterionType: Always 
```

could be used for either:
* auto approval (if it was created like this) 
* manually approval (if edited to include this after the proposal was generated)

Author: I think that having a field named `approvalCriteria` right now when we don't have support for criteria is totally misleading for users. We have the desire to have criteria and rules but we don't have a real plan now, how much it's really feasible and if it will be really useful for the community so overcommitting could not help.
With the current `autoApproval` we could have a `criteria` or `rules` subfield to match your expectation. Maybe something like this?

```yaml
autoApproval:
  criteria:
    - type: Foo
      threshold: 100
    - type: Bar
      someCriterionParameter: true
```

which anyway leaves space in the `autoApproval` field to not having just a list.

Reviewer 2: So, what about using annotation instead of changing the CRDs until we know more about the future CRD structure? E.g. if `strimzi.io/rebalance-auto-approval` is set to `true`, it is auto approved?

Author: We can go with annotation but I still don't get what makes `approvalCriteria` as list better than a generic `autoApproval` containing a field for the criteria list so allowing space for improvements. My feeling is that it's possible that we won't come back to have criteria anymore and we'll leave the annotation forever, so having an API (I talk about the `KafkaRebalance` CR) which mixes annotations and fields to provide features.

Reviewer 2: > My feeling is that it's possible that we won't come back to have criteria anymore and we'll leave the annotation forever, so having an API (I talk about the KafkaRebalance CR) which mixes annotations and fields to provide features.

Well, if we see that we don't get back to it and don't need the criteria, we can convert the annotation into a boolean flag inside the KafkaRebalance CR the future. The end result in that case will be better than the empty object.

I approved the original design, so I was fine with it. Just trying to find some compromise to move forward with this. ;-)

Author: I will do the change on the proposal from field to annotation (I see @tombentley thumbs up on this idea) and moving the current one as rejected. Should we just track somewhere (an issue? a card on the project?) that we have to come back to this in the future to evaluate if promoting to field is feasible?

Reviewer 3: > so having an API which (...) mixes annotations and fields to provide features

Well we already have that because we use an annotation for the approval anyway. If we really didn't come back to add support for criteria then leaving people with an API where the have to specify an empty object where they logically mean a boolean is pretty unintuitive. So on balance I think I prefer using an annotation for the time being until we know enough to design the API to support criteria properly.

Author: > Well we already have that because we use an annotation for the approval anyway

I see the use case different tbh. We are using annotation to take actions (approving, refreshing, ...) not for enabling or describing features. I see that the annotation usage fits pretty well when it's about "asking to the operator" to take an action on the CR. But of course, it could be just my view. I see the usage of annotation for configurability just as a temporary solution.

Author: The proposal is now updated to use an annotation.

```

> The reason for not having a simple boolean `spec.autoApproval` field is for leaving space for a future extension, maybe adding some criteria or rules that have to be matched for auto-approving the optimization proposal.

Reviewer: Ooo I like this, I can foresee some kind of pluggable `RebalanceApprover` interface.

### No auto-approval (default)

If the `spec.autoApproval` field is not specified at all, the default behavior will be the current one, so the need for manual approval by the user annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.

### Flow

As described before, the user interaction flow assumes that the `spec.autoApproval` field is specified in the `KafkaRebalance` custom resource.

1. The user creates a `KafkaRebalance` custom resource.
2. The cluster operator asks Cruise Control to generate an optimization proposal via the REST API.
3. When the optimization proposal is ready, the cluster operator checks if the user has specified the `spec.autoApproval`.
    * If the `spec.autoApproval` is not specified at all, the user has to approve the proposal manually as usual.
    * If the `spec.autoApproval` is specified, the cluster operator approves the proposal automatically by annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.

> 

Reviewer 3: What should be the current behaviour if the `autoApproval`  object (or `approvalCriteria` list if that's what we implemented) is not empty.

What should happen if it changes during the lifetime of the CR?

Author: If `autoApproval` is not empty (but it brings criteria, in the future), the KRAO should check that those criteria are met in order to approve or not the optimization proposal. Or is it anything different you were asking?

  labels:
    strimzi.io/cluster: my-cluster
spec:
  mode: # any mode
  autoApproval:
    enabled: true
```

The same result can be easily achieved if the `spec.autoApproval` is specified as empty.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:
    strimzi.io/cluster: my-cluster
spec:
  mode: # any mode
  autoApproval: {}
```

The reason for not having a simple boolean `spec.autoApproval` field is for leaving space for a future extension, maybe adding some criteria or rules that have to be matched for auto-approving the optimization proposal.

### No auto-approval (default)

If the `spec.autoApproval` field is not specified at all, the default behavior will be the current one, so the need for manual approval by the user annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.

Of course, setting the `spec.autoApproval.enabled` to `false` drives to the same result.

```yaml
apiVersion: kafka.strimzi.io/v1beta2
kind: KafkaRebalance
metadata:
  name: my-rebalance
  labels:
    strimzi.io/cluster: my-cluster
spec:
  mode: # any mode
  autoApproval:
    enabled: false

> ```

Reviewer 2: I wonder what others think about this. It might be a bit confusing to the users:
* This defaults to disabled:
  ```yaml
   spec:
     mode: ...
  ```
  => In other words, `enabled` seems to default to false.
* This defaults to enabled:
  ```yaml
   spec:
     mode: ...
     autoApproval: {}
  ```
  => Now suddenly, `enabled` seems to default to true.
* This seems to make not much sense (why would you do this if you can just leave this section out?)
  ```yaml
   spec:
     mode: ...
     autoApproval:
       enabled: false
  ```
* It would be especially weird in the future if you have something like this
  ```yaml
   spec:
     mode: ...
     autoApproval:
       enabled: false
       criteria:
         - ...
         - ...
  ```

I understand the motivation and it can be implemented in the code reasonably. I just wonder how intuitive this will be for a new user who will not read the proposal and know the background.

Would it be more natural to remove the `enabled` field completely and have it just as an empty object? I.e.
* This would disable it
  ```yaml
   spec:
     mode: ...
  ```
* This will enable it:
  ```yaml
   spec:
     mode: ...
     autoApproval: {}
  ```

I know the empty object might seem a bit weird, but if it is an future entry point, why not use it that way? And at the end, most Kafka CRs already use this in a way `userOperator: {}` => enabled with any flag.

Author: That's a really good point @scholzj, the behaviour of `enabled` is not unique in this way.
Thinking about having `autoApproval` section with `enabled: false` and maybe some criteria could be a way to "temporary" disable auto-approval? But does it really make any sense? I don't think so.
I will go with your suggestion getting rid of the `enabled` flag even because it could simplify the logic a little bit doesn't taking into account weird situations like the one just described.

### Flow

As described before, the user interaction flow assumes that the `spec.autoApproval` field is specified in the `KafkaRebalance` custom resource.

1. The user creates a `KafkaRebalance` custom resource.
2. The cluster operator asks Cruise Control to generate an optimization proposal via the REST API.
3. When the optimization proposal is ready, the cluster operator checks if the user has specified the `spec.autoApproval` and the `spec.autoApproval.enabled` is set or not:
    * If the `spec.autoApproval` is not specified at all, the user has to approve the proposal manually as usual.

>     * If the `spec.autoApproval` is specified but it's not enabled with `spec.autoApproval.enabled` set to `false`, the user has to approve the proposal manually as usual.

Reviewer 2: This does not fit with the previous chapter where you say the `enabled` will default to `true`. Maybe you wanted to say this to make it more clear?

```suggestion
    * If the `spec.autoApproval` is specified but is disabled with `spec.autoApproval.enabled` set to `false`, the user has to approve the proposal manually as usual.
```

Author: By removing the `enabled` flag, this sentence doesn't make sense anymore. I will remove it.

>     * If the `spec.autoApproval` is specified and it's enabled with `spec.autoApproval.enabled` set to `true`, the cluster operator approves the proposal automatically by annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.

Reviewer 2: ```suggestion
    * If the `spec.autoApproval` is specified and it's either enabled with `spec.autoApproval.enabled` set to `true` or when the `spec.autoApproval.enabled` field is not set at all, the cluster operator approves the proposal automatically by annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.
```

Author: It will change by removing the `enabled` flag.

>     * If the `spec.autoApproval` is specified and it's enabled with `spec.autoApproval.enabled` set to `true`, the cluster operator approves the proposal automatically by annotating the `KafkaRebalance` custom resource with the `strimzi.io/rebalance=approve` annotation.

Reviewer 2: Also ... you say `the cluster operator approves the proposal automatically by annotating the KafkaRebalance custom resource with the strimzi.io/rebalance=approve annotation`. Will it really set the annotation? Or just approve it inside the code and move it into _rebalancing_ state in the code only?

Author: In general, I would always rely on the current state machine (by applying the annotation) instead of having a specific check to run the same code for this scenario.

Reviewer 2: Ok, just wanted to make sure this is what you really wanted to do.
